asgi_correlation_id/middleware.py

```diff
@@ -1,30 +1,33 @@
 import logging
-from dataclasses import dataclass
-from uuid import UUID, uuid4
+from dataclasses import dataclass, field
+from typing import Callable, List
+from uuid import uuid4
 
 from starlette.datastructures import Headers
 from starlette.types import ASGIApp, Message, Receive, Scope, Send
 
 from asgi_correlation_id.context import correlation_id
 from asgi_correlation_id.extensions.sentry import get_sentry_extension
+from asgi_correlation_id.validators import is_valid_uuid
 
 logger = logging.getLogger('asgi_correlation_id')
 
 
-def is_valid_uuid(uuid_: str) -> bool:
-    """
-    Check whether a string is a valid v4 uuid.
-    """
-    try:
-        return bool(UUID(uuid_, version=4))
-    except ValueError:
-        return False
-
-
 @dataclass
 class CorrelationIdMiddleware:
     app: ASGIApp
     header_name: str = 'X-Request-ID'
+
+    # ID generating function
+    generator: Callable[[], str] = field(default=lambda: uuid4().hex)
+
+    # Validators for discarding badly formatted IDs
+    validators: List[Callable[[str], bool]] = None  # type: ignore[assignment]
```

Comment:
is there a better way to do this? this is what the dataclass docs said, but I don't really like it

Reply:
You should type as optional, and then none is fine

Reply:
You could use ''is_valid_uuid " as default

Reply:
Is there really a need for a list of callables here? 🤔

Reply:
No you could just expand a single validator function if you wanted to validate several things, but I'm not sure if that's the best API either. 

I guess we could use `is_valid_uuid` is a single validator callable and lets users overwrite it with another single callable. It wouldn't make a functional difference.

Reply:
Yeah, but I don't really find this intuative together with this check:
```py
elif self.validators and not all(validator(header_value) for validator in self.validators):
```
Since I'd consider the header to be valid if _any_ of the validator returns True, and next checks then can be skipped. In other words, `any` over `all`.

Reply:
If you have two validators:

- Check valid UUID formatting
- Check that it's upper-cased

How is it valid if the first one passes? You would want all validators to pass, no? My take on validators was that they would constrain what a valid request ID looks like, so I don't think using any here makes sense, unless I'm missing something

Reply:
To me, that should be one validate function. Let’s say I’d like to allow headers that start with `custom-ID-<anything>` OR a valid UUID in upper case only. Id have two validate functions, and as long as one passes, the ID is allowed through.

Reply:
I guess we can eliminate the confusion by changing `validators` to `validator` and making it a single callable like the others 🙂

Reply:
Either that, or make sure to make it very well documented. 😊